// @opentelemetry/exporter-zipkin/build/src/transform.js
"use strict";
/*
 * Copyright The OpenTelemetry Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports._toZipkinAnnotations = exports._toZipkinTags = exports.toZipkinSpan = exports.statusDescriptionTagName = exports.statusCodeTagName = void 0;
const api = require("@opentelemetry/api");
const core_1 = require("@opentelemetry/core");
const zipkinTypes = require("./types");
const ZIPKIN_SPAN_KIND_MAPPING = {
    [api.SpanKind.CLIENT]: zipkinTypes.SpanKind.CLIENT,
    [api.SpanKind.SERVER]: zipkinTypes.SpanKind.SERVER,
    [api.SpanKind.CONSUMER]: zipkinTypes.SpanKind.CONSUMER,
    [api.SpanKind.PRODUCER]: zipkinTypes.SpanKind.PRODUCER,
    // When absent, the span is local.
    [api.SpanKind.INTERNAL]: undefined,
};
exports.statusCodeTagName = 'ot.status_code';
exports.statusDescriptionTagName = 'ot.status_description';
/**
 * Translate OpenTelemetry ReadableSpan to ZipkinSpan format
 * @param span Span to be translated
 */
function toZipkinSpan(span, serviceName, statusCodeTagName, statusDescriptionTagName) {
    const zipkinSpan = {
        traceId: span.spanContext.traceId,
        parentId: span.parentSpanId,
        name: span.name,
        id: span.spanContext.spanId,
        kind: ZIPKIN_SPAN_KIND_MAPPING[span.kind],
        timestamp: core_1.hrTimeToMicroseconds(span.startTime),
        duration: core_1.hrTimeToMicroseconds(span.duration),
        localEndpoint: { serviceName },
        tags: _toZipkinTags(span.attributes, span.status, statusCodeTagName, statusDescriptionTagName, span.resource),
        annotations: span.events.length
            ? _toZipkinAnnotations(span.events)
            : undefined,
    };
    return zipkinSpan;
}
exports.toZipkinSpan = toZipkinSpan;
/** Converts OpenTelemetry Attributes and Status to Zipkin Tags format. */
function _toZipkinTags(attributes, status, statusCodeTagName, statusDescriptionTagName, resource) {
    const tags = {};
    for (const key of Object.keys(attributes)) {
        tags[key] = String(attributes[key]);
    }
    tags[statusCodeTagName] = String(api.CanonicalCode[status.code]);
    if (status.message) {
        tags[statusDescriptionTagName] = status.message;
    }
    Object.keys(resource.labels).forEach(name => (tags[name] = resource.labels[name]));
    return tags;
}
exports._toZipkinTags = _toZipkinTags;
/**
 * Converts OpenTelemetry Events to Zipkin Annotations format.
 */
function _toZipkinAnnotations(events) {
    return events.map(event => ({
        timestamp: core_1.hrTimeToMicroseconds(event.time),
        value: event.name,
    }));
}
exports._toZipkinAnnotations = _toZipkinAnnotations;
