// @opentelemetry/core/build/src/context/context.js
"use strict";
/*
 * Copyright The OpenTelemetry Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getParentSpanContext = exports.setExtractedSpanContext = exports.getExtractedSpanContext = exports.setActiveSpan = exports.getActiveSpan = exports.ACTIVE_SPAN_KEY = void 0;
const context_base_1 = require("@opentelemetry/context-base");
/**
 * Active span key
 */
exports.ACTIVE_SPAN_KEY = context_base_1.Context.createKey('OpenTelemetry Context Key ACTIVE_SPAN');
const EXTRACTED_SPAN_CONTEXT_KEY = context_base_1.Context.createKey('OpenTelemetry Context Key EXTRACTED_SPAN_CONTEXT');
/**
 * Return the active span if one exists
 *
 * @param context context to get span from
 */
function getActiveSpan(context) {
    return context.getValue(exports.ACTIVE_SPAN_KEY) || undefined;
}
exports.getActiveSpan = getActiveSpan;
/**
 * Set the active span on a context
 *
 * @param context context to use as parent
 * @param span span to set active
 */
function setActiveSpan(context, span) {
    return context.setValue(exports.ACTIVE_SPAN_KEY, span);
}
exports.setActiveSpan = setActiveSpan;
/**
 * Get the extracted span context from a context
 *
 * @param context context to get span context from
 */
function getExtractedSpanContext(context) {
    return (context.getValue(EXTRACTED_SPAN_CONTEXT_KEY) || undefined);
}
exports.getExtractedSpanContext = getExtractedSpanContext;
/**
 * Set the extracted span context on a context
 *
 * @param context context to set span context on
 * @param spanContext span context to set
 */
function setExtractedSpanContext(context, spanContext) {
    return context.setValue(EXTRACTED_SPAN_CONTEXT_KEY, spanContext);
}
exports.setExtractedSpanContext = setExtractedSpanContext;
/**
 * Get the span context of the parent span if it exists,
 * or the extracted span context if there is no active
 * span.
 *
 * @param context context to get values from
 */
function getParentSpanContext(context) {
    var _a;
    return ((_a = getActiveSpan(context)) === null || _a === void 0 ? void 0 : _a.context()) || getExtractedSpanContext(context);
}
exports.getParentSpanContext = getParentSpanContext;
